c2rust-analyze/src/c_void_casts.rs

```diff
@@ -0,0 +1,368 @@
+    ///
+    /// [`Assign`]: rustc_middle::mir::StatementKind::Assign
+    /// [`Cast`]: rustc_middle::mir::Rvalue::Cast
+    /// [`From`]: CVoidCastDirection::From
+    /// [`To`]: CVoidCastDirection::To
+    pub fn should_skip_stmt(&self, stmt: &Statement<'tcx>) -> bool {
+        let (lhs, rv) = match get_assign_sides(stmt) {
+            None => return false,
+            Some(it) => it,
+        };
+        self.to.contains(lhs)
+            || self.from.contains(match get_cast_place(rv) {
+                None => return false,
+                Some(it) => it,
+            })
+    }
+
+    /// Insert all applicable [`*c_void`] casts
+    /// from a function [`Body`] into this [`CVoidCasts`].
+    ///
+    /// That is, find all calls to:
+    ///
+    /// * `malloc`
+    /// * `calloc`
+    /// * `realloc`
+    /// * `free`
+    ///
+    /// and insert their casts to and from [`*c_void`].
+    ///
+    /// Note that since [`Place`]s are local to a function [`Body`],
+    /// a [`CVoidCasts`] is only valid when it contains things from one [`Body`].
+    ///
+    /// [`*c_void`]: core::ffi::c_void
+    fn insert_all_from_body(&mut self, body: &Body<'tcx>, tcx: TyCtxt<'tcx>) {
+        for bb_data in body.basic_blocks().iter() {
+            let term: &Terminator = match &bb_data.terminator {
+                Some(term) => term,
+                None => continue,
+            };
+            let (func, args, destination, target) = match term.kind {
+                TerminatorKind::Call {
+                    ref func,
+                    ref args,
+                    destination,
+                    target,
+                    ..
+                } => (func, args, destination, target),
+                _ => continue,
+            };
+            let func_ty = func.ty(&body.local_decls, tcx);
+
+            // For [`CVoidCastDirection::From`], we only count
+            // a cast from `*c_void` to an arbitrary type in the subsequent block,
+            // searching forward.
+            let find_first_cast_succ_block = |get_cast| {
+                let successor_block_id = target.unwrap();
+                body.basic_blocks()[successor_block_id]
+                    .statements
+                    .iter()
+                    .find_map(get_cast)
+            };
+
+            // For [`CVoidCastDirection::From`], we only count
+            // a cast to `*c_void` from an arbitrary type in the same block,
+            // searching backwards.
+            let find_last_cast_curr_block =
+                |get_cast| bb_data.statements.iter().rev().find_map(get_cast);
+
+            for direction in CVoidCastDirection::from_callee(ty_callee(tcx, func_ty))
+                .iter()
+                .copied()
+            {
+                use CVoidCastDirection::*;
+                let c_void_ptr = match direction {
+                    From => destination,
+                    To => args[0].place().unwrap(),
```

Comment:
This `unwrap` will fail on the following code:
```Rust
use libc::{free, c_void};
const NULL: *mut c_void = 0 as *mut _;
unsafe fn f() {
    free(NULL);
}
```
The MIR for this has an `Operand::Constant` for the argument to `free`.

Reply:
If you're okay with it I'd like to make it an `expect` instead and address the case you outlined in the same PR as the one where we support null pointers

Reply:
Seems reasonable

Reply:
9c17f600